packages/ra-core/src/store/useStore.ts

```diff
@@ -46,25 +47,62 @@ import { useStoreContext } from './useStoreContext';
  */
 export const useStore = <T = any>(
     key: string,
-    defaultValue?: T
-): useStoreResult<T> => {
+    defaultValue?: T,
+    validate?: ValidateStoreValue<T>
+): UseStoreResult<T> => {
     const { getItem, setItem, subscribe } = useStoreContext();
+    if (
+        typeof validate === 'function' &&
+        typeof defaultValue !== 'undefined' &&
+        !validate(defaultValue)
+    ) {
+        console.warn(`Invalid default value for store key ${key}`);
+    }
     const [value, setValue] = useState(() => getItem(key, defaultValue));
 
+    useEffect(() => {
```

Comment:
In the following scenario, this will not work:

1. A user calls setStore with an invalid value
2. the state setter updates the state
3. the components relying on that state rerender, possibly with an invalid state shape
4. the effect runs, logs a warning and sets the value back to default. 

Too late: the error already occured in step 3

Besides, you already check validation in the set method, I don't see any case where this effect is necessary.

Reply:
The idea was to validate the initial value from the store but you're right, it's incorrect